core/app/models/spree/promotion/rules/user_role.rb

```diff
@@ -0,0 +1,26 @@
+module Spree
+  class Promotion
+    module Rules
+      class UserRole < PromotionRule
+        has_many :promotion_rule_roles,
+          class_name: 'Spree::PromotionRuleRole',
+          foreign_key: :promotion_rule_id,
+          dependent: :destroy
+
+        has_many :roles,
+          through: :promotion_rule_roles,
+          class_name: 'Spree::Role'
+
+        accepts_nested_attributes_for :roles
+
+        def applicable?(promotable)
+          promotable.is_a?(Spree::Order)
+        end
+
+        def eligible?(order, options = {})
+          (roles & order.user.spree_roles).any? if order.user
```

Comment:
at scale (like in the case of abuse by the Marketing team, as described below), this will not be performant because the filtering will be done in Ruby. For example, let's say if there were hundreds or thousands of roles attached to a single User, this will cause significant bloat in the app (so much bloat it would cause a checkout to completely to a halt)

you could actually work-around that scalability problem by passing the logic deep into the SQL query itself, like so:

```order.user.role_users.where(role_id: roles.collect(:id))```

this would make it more performant and remove the need for AR to load all of the user's roles at once, thus instantiating too many AR objects.